yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -290,8 +290,18 @@ private[yarn] class YarnAllocator(
       location: String,
       containersToUse: ArrayBuffer[Container],
       remaining: ArrayBuffer[Container]): Unit = {
+    // SPARK-6050: certain Yarn configurations return a resource configuration that doesn't match
+    // the request; for example, capacity scheduler + DefaultResourceCalculator. Allow users in
+    // those situations to disable resource matching.
+    val matchingResource =
+        if (sparkConf.getBoolean("spark.yarn.container.matchAnyResource", false)) {
```

Comment:
Take a look at the latest version to see if it's any clearer.

But the gist is that `amClient.getMatchingResources()` is matching the resources you asked for (which is `resource`) against the parameter you're passing. What the option controls is whether you're passing `resource` also as the resource to match against - so basically, the exact same structure that is already in the outstanding list of requests.

So I think you're reading the condition backwards. Or something.